### DISCUSSION TITLE
How to change state of a StateProvider outside a widget?

### DISCUSSION THREAD

**Discussion by ramtinir** [rrousselGit/riverpod#1648] [Q&A]
Suppose we have this in the global scope:

`final loginStateProvider = StateProvider((ref) => LoginState.initial);`

and there's also:

```
class LoginFormNotifier extends StateNotifier<LoginInputsState> {
  LoginFormNotifier()
      : super(const LoginFormState(
            phone: '',
            password: '');

  void setPhone(String phone) {
    if (phone == '') {
        ///// !!! I want to set loginStateProvider's state to another value here, something like:

              ref.read(loginStateProvider.notifier).state = LoginState.error;
```

But we don't have access to `ref` here. How is it possible to change the state of the `loginStateProvider` here?

**Comment by rrousselGit** [comment]
You have a StateNotifier here. So chances are you have a StateNotifierProvider too (if not, you probably should).

Considering that, you can pass the ref from your StateNotifierProvider to your StateNotifier (in its constructor). Then `setPhone` should be able to use that `ref` to read providers.
